File: api/src/main/java/com/tainzhi/sample/api/adapter/BasicVertialAdapter.java
package com.tainzhi.sample.api.adapter;

import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.chad.library.adapter.base.BaseQuickAdapter;
import com.chad.library.adapter.base.BaseViewHolder;
import com.tainzhi.sample.api.R;

/**
 * @author: tainzhi
 * @mail: qfq61@qq.com
 * @date: 2019-10-09 08:50
 * @description:
 **/

public class BasicVertialAdapter extends BaseQuickAdapter<String, BaseViewHolder> {
	public boolean isWaterFall = false;
	
	public BasicVertialAdapter(boolean isWaterFall) {
		super(R.layout.item_basic_vertial_view);
		this.isWaterFall = isWaterFall;
	}
	
	@Override
	protected void convert(@NonNull BaseViewHolder helper, String item) {
		// 瀑布流式布局
		if (isWaterFall) {
			View view = helper.itemView;
			ViewGroup.LayoutParams params = view.getLayoutParams();
			int newHeight = getRandomHeight(params.height);
			params.height = newHeight;
			view.setLayoutParams(params);
			helper.setText(R.id.text_view, item);
		} else {
			helper.setText(R.id.text_view, item);
		}
	}
	
	private int getRandomHeight(double oriheight) {
		double height = oriheight + oriheight * Math.random();
		if (height > 1300) {
			height = 300;
		}
		return (int) height;
	}
	
	@Override
	public void onAttachedToRecyclerView(@NonNull RecyclerView recyclerView) {
		super.onAttachedToRecyclerView(recyclerView);
		RecyclerView.LayoutManager manager = recyclerView.getLayoutManager();
		if (manager instanceof GridLayoutManager) {
			final GridLayoutManager layoutManager = ((GridLayoutManager)manager);
			layoutManager.setSpanSizeLookup(new GridLayoutManager.SpanSizeLookup() {
				@Override
				public int getSpanSize(int position) {
					// 模 9 为 0 的 item, 占位为整个屏幕 2/2
					// 模 9 为非 1 的 item, 占位为整个屏幕的 1/2
					return (position % 9) == 0 ? 2 : 1;
				}
			});
		}
	}
	
}

File: module-android-api/src/main/java/com/tainzhi/sample/api/adapter/BasicVertialAdapter.kt
package com.tainzhi.sample.api.adapter

import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.GridLayoutManager.SpanSizeLookup
import androidx.recyclerview.widget.RecyclerView
import com.chad.library.adapter.base.BaseQuickAdapter
import com.chad.library.adapter.base.viewholder.BaseViewHolder
import com.tainzhi.sample.api.R

/**
 * @author: tainzhi
 * @mail: qfq61@qq.com
 * @date: 2019-10-09 08:50
 * @description:
 */
class BasicVertialAdapter(isWaterFall: Boolean) :
    BaseQuickAdapter<String?, BaseViewHolder>(R.layout.item_basic_vertial_view) {
    var isWaterFall = false
    override fun convert(helper: BaseViewHolder, item: String?) {
        // 瀑布流式布局
        if (isWaterFall) {
            val view = helper.itemView
            val params = view.layoutParams
            val newHeight = getRandomHeight(params.height.toDouble())
            params.height = newHeight
            view.layoutParams = params
            helper.setText(R.id.text_view, item)
        } else {
            helper.setText(R.id.text_view, item)
        }
    }

    private fun getRandomHeight(oriheight: Double): Int {
        var height = oriheight + oriheight * Math.random()
        if (height > 1300) {
            height = 300.0
        }
        return height.toInt()
    }

    override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
        super.onAttachedToRecyclerView(recyclerView)
        val manager = recyclerView.layoutManager
        if (manager is GridLayoutManager) {
            manager.spanSizeLookup = object : SpanSizeLookup() {
                override fun getSpanSize(position: Int): Int {
                    // 模 9 为 0 的 item, 占位为整个屏幕 2/2
                    // 模 9 为非 1 的 item, 占位为整个屏幕的 1/2
                    return if (position % 9 == 0) 2 else 1
                }
            }
        }
    }

    init {
        this.isWaterFall = isWaterFall
    }
}